Fix #7847: Use ViewportSign coordinates for sign Kdtree coordinates

nielsmh opened the pull request:
Fix #7847 by ensuring the same coordinates are used for a station sign regardless of how the landscape changes below it after the coordinates were first determined.

By keeping track of whether the ViewportSign is valid for Kdtree use (and only ever registering the station viewport sign when the sign object is valid) a lot of code can be simplified, and it should prevent a lot of similar bugs too.

I think this is the real solution to all Kdtree bugs regarding station signs. I think something similar needs to be done for town names and player signs too.

nielsmh committed:
Change: Use same sign position calculation in Station::UpdateVirtCoord() and ViewportSignKdtreeItem::MakeStation()

nielsmh committed:
Fix #7847: Use ViewportSign coordinates for sign Kdtree coordinates for stations

Keeping track of whether the ViewportSign data are valid for Kdtree usage allows moving more housekeeping to Station::UpdateVirtCoord and Waypoint::UpdateVirtCoord and helps simplify a lot of other code.

SamuXarick commented on src/waypoint_cmd.cpp:
```diff
@@ -39,8 +39,13 @@
 void Waypoint::UpdateVirtCoord()
 {
 	Point pt = RemapCoords2(TileX(this->xy) * TILE_SIZE, TileY(this->xy) * TILE_SIZE);
```
I think you forgot to change this line.

nielsmh replied:
That's the old, more correct method. Changing the one for regular stations back instead.

nielsmh committed:
Codechange: Rename StationViewportSign to TrackedViewportSign and move it to viewport_type.h

nielsmh committed:
Fix: Track validity of player signs for Kdtree via TrackedViewportSign

nielsmh committed:
Fix: Track validity of town signs for Kdtree via TrackedViewportSign

nielsmh committed:
Change: Go back to using RemapCoords2 for station/waypoint sign positions

It's the more correct approach, it was only not usable before due to inconsistencies with when sign positions were updated.

nielsmh commented:
Tested #7371 report, PSG 201 save remains loadable without crashing.

Tested #7440 report, is still able to create and then remove a town in scenario editor.

Tested #7481 and #7530 via the save provided in the latter, seems to pass the oil rig deletion.

glx22 reviewed (approved):
Looks good to me
